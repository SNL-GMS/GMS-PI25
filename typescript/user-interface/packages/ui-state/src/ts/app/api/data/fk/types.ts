/**
 * Input type for Compute FK service call. This input
 * is compatible with the COI input i.e. start/end are strings
 */
import type { FkTypes } from '@gms/common-model';

import type { AsyncFetchHistory } from '../../../query';

/**
 * Defines the history record type for the getChannelSegmentsByChannel query
 */
export type ComputeFkSpectraHistory = AsyncFetchHistory<FkTypes.FkInputWithConfiguration>;
