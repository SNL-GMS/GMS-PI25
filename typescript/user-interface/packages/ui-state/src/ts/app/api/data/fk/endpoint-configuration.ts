import { FkControlUrls } from '@gms/common-model/lib/endpoints/types';
import { UI_URL } from '@gms/common-util';
import type { RequestConfig, ServiceDefinition } from '@gms/ui-workers';

import { prioritizeRequests } from '../../request-priority';

/**
 * Compute Fk request config definition
 */
export interface FkControlRequestConfig extends RequestConfig {
  readonly computeFkSpectra: {
    readonly baseUrl: string;
    readonly services: {
      readonly computeFkSpectra: ServiceDefinition;
    };
  };
}

const baseUrl = `${UI_URL}${FkControlUrls.baseUrl}`;

/**
 * The Waveform request config for all services.
 */
export const config: FkControlRequestConfig = {
  computeFkSpectra: {
    baseUrl,
    // Service endpoints for this component
    services: prioritizeRequests({
      computeFkSpectra: {
        friendlyName: FkControlUrls.computeFkSpectra.friendlyName,
        requestConfig: {
          baseURL: baseUrl,
          method: 'post',
          url: FkControlUrls.computeFkSpectra.url,
          proxy: false,
          headers: {
            // configure to receive msgpack encoded data
            accept: 'application/msgpack',
            'content-type': 'application/json'
          },
          timeout: 180000 // 3 mins
        }
      }
    })
  }
};

export type FkComputeServices = keyof typeof config.computeFkSpectra.services;
