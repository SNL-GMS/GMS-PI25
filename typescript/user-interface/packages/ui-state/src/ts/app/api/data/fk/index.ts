export * from './compute-fk-spectra';
export * from './mark-fk-reviewed';
export * from './mutate-fk-channel-segment-record';
export * from './types';
