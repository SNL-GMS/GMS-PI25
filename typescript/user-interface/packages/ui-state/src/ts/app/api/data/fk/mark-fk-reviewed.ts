import type { ChannelSegmentTypes } from '@gms/common-model';
import type { CaseReducer } from '@reduxjs/toolkit';
import { createAction } from '@reduxjs/toolkit';

import { createChannelSegmentString } from '../../../../workers/waveform-worker/util/channel-segment-util';
import type { DataState } from '../types';

/**
 * The action for marking the FK as reviewed.
 */
export const markFkReviewed = createAction<
  {
    channelSegmentDescriptor: ChannelSegmentTypes.ChannelSegmentDescriptor;
  },
  'data/markFkReviewed'
>('data/markFkReviewed');

/**
 * Mutates a writable draft fk channel segment record with new fk channel segment
 *
 * @param state the current redux state of the slice
 * @param action the action being invoked
 */
export const markFkReviewedReducer: CaseReducer<DataState, ReturnType<typeof markFkReviewed>> = (
  state,
  action
) => {
  const { channelSegmentDescriptor: fkChannelSegmentDescriptor } = action.payload;
  const fkChannelSegmentRecord = state.fkChannelSegments;

  // Update the FkChannelSegmentRecord timeseries (fk) as reviewed
  const channelDescriptorString = createChannelSegmentString(fkChannelSegmentDescriptor);
  fkChannelSegmentRecord[channelDescriptorString].timeseries.forEach(ts => {
    // eslint-disable-next-line no-param-reassign
    ts.reviewed = true;
  });
};
