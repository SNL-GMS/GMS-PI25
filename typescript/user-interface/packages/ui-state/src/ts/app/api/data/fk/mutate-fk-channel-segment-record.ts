import type { ChannelSegmentTypes, FkTypes } from '@gms/common-model';
import { SignalDetectionTypes } from '@gms/common-model';
import type { WritableDraft } from 'immer/dist/internal';

import type {
  FkChannelSegmentRecord,
  FkFrequencyThumbnailRecord,
  SignalDetectionsRecord
} from '../../../../types';
import { createChannelSegmentString } from '../../../../workers/waveform-worker/util/channel-segment-util';

/**
 * Mutates a writable draft fk channel segment record with new fk channel segment
 * ! Mutates the draft in place
 *
 * @param FkChannelSegmentRecord the  fk channel segments
 * @param fkChannelSegment the fk channel segment
 */
export const mutateFkChannelSegmentsRecord = (
  draft: WritableDraft<FkChannelSegmentRecord>,
  fkChannelSegment: ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>
): void => {
  if (!fkChannelSegment) return;

  // Update the FkChannelSegmentRecord
  const channelDescriptorString = createChannelSegmentString(fkChannelSegment.id);
  draft[channelDescriptorString] = fkChannelSegment;
};

/**
 * Update FkFrequencyThumbnailRecord in state with frequency thumbnail
 *
 * @param args FkInputWithConfiguration
 * @param draft FkFrequencyThumbnailRecord
 * @param fkChannelSegment fk channel segment returned from computeFk call
 */
export const mutateFkThumbnailRecord = (
  args: FkTypes.FkInputWithConfiguration,
  draft: WritableDraft<FkFrequencyThumbnailRecord>,
  fkChannelSegment: ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>
): void => {
  if (!fkChannelSegment || !args) return;

  if (!draft[args.signalDetectionId]) {
    draft[args.signalDetectionId] = [];
  }

  // Find the index of thumbnail to replace else add it
  const frequencyBand: FkTypes.FrequencyBand = {
    minFrequencyHz: args.fkComputeInput.lowFrequency,
    maxFrequencyHz: args.fkComputeInput.highFrequency
  };
  const frequencyFk: FkTypes.FkFrequencyThumbnail = {
    fkSpectra: fkChannelSegment.timeseries[0],
    frequencyBand
  };
  const index = draft[args.signalDetectionId].findIndex(
    thumbnail =>
      thumbnail.frequencyBand.minFrequencyHz === frequencyBand.minFrequencyHz &&
      thumbnail.frequencyBand.maxFrequencyHz === frequencyBand.maxFrequencyHz
  );
  // replace
  if (index >= 0) {
    draft[args.signalDetectionId][index] = frequencyFk;
  } else {
    draft[args.signalDetectionId].push(frequencyFk);
  }
};

/**
 * TODO: Need to integrate into SD editing work
 * Update the Azimuth FM channel segment descriptor to fkChannelSegment
 *
 * @param draft
 * @param signalDetectionId
 * @param fkChannelSegment
 * @returns
 */
export const mutateSignalDetectionRecord = (
  draftSdRecord: WritableDraft<SignalDetectionsRecord>,
  signalDetectionId: string,
  fkChannelSegment: ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>
): void => {
  if (!fkChannelSegment || !draftSdRecord || !draftSdRecord[signalDetectionId]) return;

  // Update Signal Detection Azimuth FM with ChannelSegment descriptor from FK
  const draftHypos: WritableDraft<SignalDetectionTypes.SignalDetectionHypothesis> = SignalDetectionTypes.Util.getCurrentHypothesis(
    draftSdRecord[signalDetectionId].signalDetectionHypotheses
  );
  draftHypos.featureMeasurements.forEach(draft => {
    if (
      draft.featureMeasurementType ===
        SignalDetectionTypes.FeatureMeasurementType.RECEIVER_TO_SOURCE_AZIMUTH ||
      draft.featureMeasurementType ===
        SignalDetectionTypes.FeatureMeasurementType.SOURCE_TO_RECEIVER_AZIMUTH
    ) {
      draft.analysisWaveform.waveform.id = fkChannelSegment.id;
    }
  });
};
