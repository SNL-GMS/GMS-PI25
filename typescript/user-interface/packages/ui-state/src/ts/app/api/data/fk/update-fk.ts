import type { WaveformTypes } from '@gms/common-model';
import { ChannelSegmentTypes, FkTypes } from '@gms/common-model';

/**
 * Helper method to create the FkData waveforms (azimuthWf, fstatWf, slownessWf)
 *
 * @param fkSpectra the fk spectra
 */
function createFkWaveform(fkSpectra: FkTypes.FkPowerSpectraCOI): WaveformTypes.Waveform {
  return {
    sampleRateHz: fkSpectra.sampleRateHz,
    sampleCount: fkSpectra.sampleCount,
    startTime: fkSpectra.startTime,
    endTime: fkSpectra.endTime,
    type: ChannelSegmentTypes.TimeSeriesType.FK_SPECTRA,
    samples: []
  };
}

/**
 * Convert a FkSpectra (received from COI or Streaming Service) into an FstatData representation.
 *
 * @param fkSpectra: FkPowerSpectra from COI/Streaming Service
 * @param beamWaveform: beam from the SD Arrival Time Feature measurement Channel Segment
 * @param arrivalTime: arrival time value
 *
 * @returns FK Stat Data or undefined if not able to create
 */
function convertToPlotData(fkSpectra: FkTypes.FkPowerSpectraCOI): FkTypes.FstatData | undefined {
  // If the channel segment is populated at the top properly
  if (!fkSpectra) {
    return undefined;
  }

  const fstatData: FkTypes.FstatData = {
    azimuthWf: createFkWaveform(fkSpectra),
    fstatWf: createFkWaveform(fkSpectra),
    slownessWf: createFkWaveform(fkSpectra)
  };

  // Populate fstatData waveforms beams was a parameter
  if (fkSpectra && fkSpectra.values) {
    fkSpectra.values.forEach((fkSpectrum: FkTypes.FkPowerSpectrum) => {
      fkSpectrum.attributes.forEach(attributes => {
        fstatData.azimuthWf.samples.push(attributes.azimuth);
        fstatData.fstatWf.samples.push(attributes.peakFStat);
        fstatData.slownessWf.samples.push(attributes.slowness);
      });
    });
  }
  return fstatData;
}

/**
 * Update the COI FK Power Spectra to include missing UI fields
 *
 * @param fk COI FK
 * @param args FkInputWithConfiguration (includes Configuration to restore)
 * @returns FkPowerSpectra (UI version)
 */
export const updateFk = (
  fk: FkTypes.FkPowerSpectra,
  args: FkTypes.FkInputWithConfiguration
): FkTypes.FkPowerSpectra => {
  if (!fk) {
    return undefined;
  }
  return {
    ...fk,
    fstatData: convertToPlotData(fk as FkTypes.FkPowerSpectraCOI),
    configuration: args.configuration,
    reviewed: false,
    // These should be coming back from the backend?
    windowLead: args.windowParams.leadSeconds,
    windowLength: args.windowParams.lengthSeconds,
    stepSize: args.windowParams.stepSize,
    lowFrequency: args.fkComputeInput.lowFrequency,
    highFrequency: args.fkComputeInput.highFrequency,
    metadata: {
      ...fk.metadata,
      slowStartX: FkTypes.Util.degreeToKmApproximate(fk.metadata.slowStartX),
      slowStartY: FkTypes.Util.degreeToKmApproximate(fk.metadata.slowStartY),
      slowDeltaX: FkTypes.Util.degreeToKmApproximate(fk.metadata.slowDeltaX),
      slowDeltaY: FkTypes.Util.degreeToKmApproximate(fk.metadata.slowDeltaY)
    }
  };
};
