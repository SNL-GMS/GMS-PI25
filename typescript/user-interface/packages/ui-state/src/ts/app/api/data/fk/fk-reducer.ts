import type { ActionReducerMapBuilder } from '@reduxjs/toolkit';

import type { DataState } from '../types';
import { markFkReviewed, markFkReviewedReducer } from './mark-fk-reviewed';

/**
 * Injects the FK reducers to the provided builder.
 *
 * @param builder the action reducer map builder
 */
export const addFkReducers = (builder: ActionReducerMapBuilder<DataState>): void => {
  builder.addCase(markFkReviewed, markFkReviewedReducer);
};
