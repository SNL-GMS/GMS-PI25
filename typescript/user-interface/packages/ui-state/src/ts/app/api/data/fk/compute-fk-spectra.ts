import type { ChannelSegmentTypes, FkTypes } from '@gms/common-model';
import { UILogger } from '@gms/ui-util';
import type { ActionReducerMapBuilder } from '@reduxjs/toolkit';
import { createAsyncThunk } from '@reduxjs/toolkit';
import produce from 'immer';

import type { AppState, DataState } from '../../../../ui-state';
import { computeFkSpectraWorker } from '../../../../workers';
import { AsyncActionStatus } from '../../../query';
import { hasAlreadyBeenRequested } from '../../../query/async-fetch-util';
import { config } from './endpoint-configuration';
import {
  mutateFkChannelSegmentsRecord,
  mutateFkThumbnailRecord,
  mutateSignalDetectionRecord
} from './mutate-fk-channel-segment-record';
import { updateFk } from './update-fk';

const logger = UILogger.create(
  'GMS_LOG_COMPUTE_FK_SPECTRA',
  process.env.GMS_LOG_COMPUTE_FK_SPECTRA
);

const createRequestString = (args: FkTypes.FkInputWithConfiguration): string => {
  return `${args.signalDetectionId}.${JSON.stringify(args.fkComputeInput)}`;
};
/**
 * Helper function used to determine if the computeFkSpectra query should be skipped.
 *
 * @returns returns true if the arguments are valid; false otherwise.
 */
export const shouldSkipComputeFkSpectra = (args: FkTypes.FkInputWithConfiguration): boolean =>
  !args ||
  args.configuration == null ||
  args.fkComputeInput == null ||
  args.signalDetectionId == null;

/**
 * Async thunk action that request a FK be computed.
 */
export const computeFkSpectra = createAsyncThunk<
  ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>,
  FkTypes.FkInputWithConfiguration
>(
  'fk/computeFkSpectra',
  // eslint-disable-next-line @typescript-eslint/require-await
  async (arg: FkTypes.FkInputWithConfiguration, { rejectWithValue }) => {
    const requestConfig = {
      ...config.computeFkSpectra.services.computeFkSpectra.requestConfig,
      data: arg.fkComputeInput
    };
    return computeFkSpectraWorker(requestConfig).catch(error => {
      if (error.message !== 'canceled') {
        logger.error(`Failed computeFkSpectra (rejected)`, error);
      }
      return rejectWithValue(error);
    });
  },
  {
    condition: (arg: FkTypes.FkInputWithConfiguration, { getState }) => {
      const state = (getState as () => AppState)();

      // determine if the query should be skipped based on the provided args; check if valid
      if (shouldSkipComputeFkSpectra(arg)) {
        return false;
      }

      // check if the query has been executed already
      const requests = state.data.queries.computeFkSpectra[createRequestString(arg)] ?? {};
      return !hasAlreadyBeenRequested<FkTypes.FkInputWithConfiguration>(requests, arg);
    }
  }
);

/**
 * Injects the computeFkSpectra reducers to the provided builder.
 *
 * @param builder the action reducer map builder
 */
export const addComputeFkSpectraReducers = (builder: ActionReducerMapBuilder<DataState>): void => {
  builder
    /**
     * computeFkSpectra PENDING action
     * Updates the signal detection query state to indicate that the query status is pending.
     * Note: Mutating the state maintains immutability because it uses immer under the hood.
     */
    .addCase(computeFkSpectra.pending, (state, action) => {
      const history = state.queries.computeFkSpectra;
      if (!history[createRequestString(action.meta.arg)]) {
        history[createRequestString(action.meta.arg)] = {};
      }
      history[createRequestString(action.meta.arg)][action.meta.requestId] = {
        arg: action.meta.arg,
        status: AsyncActionStatus.pending,
        error: undefined
      };
    })

    /**
     * computeFkSpectra FULFILLED action
     * Updates the signal detection query state to indicate that the query status is fulfilled.
     * Stores the retrieved signal detections in the signal detection redux state.
     * Stores the retrieved channel segments in the channel segment redux state.
     * Note: Mutating the state maintains immutability because it uses immer under the hood.
     */
    .addCase(computeFkSpectra.fulfilled, (state, action) => {
      const history = state.queries.computeFkSpectra;
      // make sure this is set in case a signal detection is returned from the worker after the
      // interval has changed
      if (!history[createRequestString(action.meta.arg)]) {
        history[createRequestString(action.meta.arg)] = {};
      }
      // If we don't have a request matching this ID, that means that it was cleared out
      // (for example, when an interval is closed), and so we don't need to process this.
      if (
        !Object.hasOwnProperty.call(
          history[createRequestString(action.meta.arg)],
          action.meta.requestId
        )
      ) {
        return;
      }
      history[createRequestString(action.meta.arg)][action.meta.requestId] = {
        arg: action.meta.arg,
        status: AsyncActionStatus.fulfilled,
        error: undefined
      };

      // Restore config and process waveforms
      const fkResult: ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra> = action.payload;
      const fk = produce(fkResult, draft => {
        draft.timeseries = draft.timeseries.map(tsFk => {
          return updateFk(tsFk, action.meta.arg);
        });
      });
      if (!action.meta.arg.isThumbnailRequest) {
        mutateFkChannelSegmentsRecord(state.fkChannelSegments, fk);
        mutateSignalDetectionRecord(state.signalDetections, action.meta.arg.signalDetectionId, fk);
      } else {
        mutateFkThumbnailRecord(action.meta.arg, state.fkFrequencyThumbnails, fk);
      }
    })

    /**
     * computeFkSpectra REJECTED action
     * Updates the signal detection query state to indicate that the query status is rejected,
     * and adds the error message.
     * Note: Mutating the state maintains immutability because it uses immer under the hood.
     */
    .addCase(computeFkSpectra.rejected, (state, action) => {
      const history = state.queries.computeFkSpectra;
      // don't update if the history has been cleared before this promise rejected
      if (!Object.hasOwnProperty.call(history, createRequestString(action.meta.arg))) {
        return;
      }
      history[createRequestString(action.meta.arg)][action.meta.requestId] = {
        arg: action.meta.arg,
        status: AsyncActionStatus.rejected,
        error: action.error
      };
    });
};
