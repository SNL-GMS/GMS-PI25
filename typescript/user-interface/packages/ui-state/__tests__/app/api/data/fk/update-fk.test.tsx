/* eslint-disable @typescript-eslint/no-magic-numbers */
import type { FkTypes } from '@gms/common-model';

import { updateFk } from '../../../../../src/ts/app/api/data/fk/update-fk';
import { fkInput, getTestFkCoiData } from '../../../../__data__';

const fkSpectraCoi = getTestFkCoiData(2000);
describe('Update Fk after computeFkSpectra call', () => {
  it('update Fk COI', () => {
    const result = updateFk(fkSpectraCoi as FkTypes.FkPowerSpectra, fkInput);
    const resultWithoutValues = {
      ...result,
      values: []
    };
    expect(resultWithoutValues).toMatchSnapshot();
  });

  it('update Fk COI can handle undefined', () => {
    expect(updateFk(undefined, fkInput)).toMatchSnapshot();
  });
});
