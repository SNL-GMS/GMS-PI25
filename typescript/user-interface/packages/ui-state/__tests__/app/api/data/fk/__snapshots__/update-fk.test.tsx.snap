// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Fk after computeFkSpectra call update Fk COI 1`] = `
{
  "configuration": {
    "contributingChannelsConfiguration": [
      {
        "enabled": true,
        "id": "BRTR.BR101.SHZ",
        "name": "BRTR.BR101.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR102.SHZ",
        "name": "BRTR.BR102.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR103.SHZ",
        "name": "BRTR.BR103.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR104.SHZ",
        "name": "BRTR.BR104.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR105.SHZ",
        "name": "BRTR.BR105.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR106.SHZ",
        "name": "BRTR.BR106.SHZ",
      },
      {
        "enabled": true,
        "id": "BRTR.BR131.BHE",
        "name": "BRTR.BR131.BHE",
      },
      {
        "enabled": true,
        "id": "BRTR.BR131.BHN",
        "name": "BRTR.BR131.BHN",
      },
      {
        "enabled": true,
        "id": "BRTR.BR131.BHZ",
        "name": "BRTR.BR131.BHZ",
      },
    ],
    "leadFkSpectrumSeconds": 1,
    "maximumSlowness": 40,
    "mediumVelocity": 5.8,
    "normalizeWaveforms": false,
    "numberOfPoints": 81,
    "useChannelVerticalOffset": false,
  },
  "contribChannels": [],
  "endTime": 2302,
  "fstatData": {
    "azimuthWf": {
      "endTime": 2302,
      "sampleCount": 151,
      "sampleRateHz": 0.5,
      "samples": [
        322.431407971,
        217.874983651,
        285.255118703,
        197.102728969,
        194.931417178,
        235.007979801,
        315,
        259.380344724,
        280.304846469,
        233.130102354,
        160.559965172,
        278.746162263,
        254.054604099,
        236.309932474,
        244.798876355,
        270,
        235.619655276,
        221.987212496,
        244.798876355,
        263.157226587,
        210.465544919,
        229.763641691,
        167.905242923,
        188.746162263,
        246.801409486,
        169.992020199,
        254.054604099,
        218.659808254,
        251.565051177,
        284.036243468,
        270,
        288.434948823,
        296.565051177,
        284.036243468,
        270,
        270,
        288.434948823,
        288.434948823,
        288.434948823,
        296.565051177,
        303.690067526,
        225,
        225,
        230.194428908,
        26.5650511771,
        308.290163192,
        313.025065989,
        285.945395901,
        217.405356631,
        344.054604099,
        296.565051177,
        180,
        323.615648184,
        305.537677792,
        281.309932474,
        225,
        315,
        315,
        296.565051177,
        108.434948823,
        161.565051177,
        284.036243468,
        291.370622269,
        14.0362434679,
        157.520565603,
        243.434948823,
        129.805571092,
        180,
        297.758540601,
        33.690067526,
        14.0362434679,
        194.036243468,
        293.749494493,
        26.5650511771,
        296.565051177,
        355.236358309,
        85.0302592719,
        297.897271031,
        330.255118703,
        0,
        291.801409486,
        180,
        345.963756532,
        344.054604099,
        354.289406863,
        180,
        255.068582822,
        45,
        168.690067526,
        347.471192291,
        206.565051177,
        284.036243468,
        251.565051177,
        213.690067526,
        190.304846469,
        28.610459666,
        170.134193057,
        270,
        329.036243468,
        180,
        243.434948823,
        182.726310994,
        261.253837737,
        218.659808254,
        230.194428908,
        159.67686317,
        215.537677792,
        246.801409486,
        254.475889003,
        282.094757077,
        281.309932474,
        281.309932474,
        270,
        284.036243468,
        255.963756532,
        228.814074834,
        326.309932474,
        146.309932474,
        296.565051177,
        206.565051177,
        197.102728969,
        262.874983651,
        255.963756532,
        255.963756532,
        236.309932474,
        254.054604099,
        261.869897646,
        251.565051177,
        293.198590514,
        278.130102354,
        308.659808254,
        203.198590514,
        270,
        306.869897646,
        255.963756532,
        206.565051177,
        284.036243468,
        258.690067526,
        209.744881297,
        234.462322208,
        307.303948278,
        264.289406863,
        213.690067526,
        285.945395901,
        305.537677792,
        288.434948823,
        258.690067526,
        310.601294645,
        239.036243468,
        240.642246457,
        246.801409486,
      ],
      "startTime": 2000,
      "type": "FK_SPECTRA",
    },
    "fstatWf": {
      "endTime": 2302,
      "sampleCount": 151,
      "sampleRateHz": 0.5,
      "samples": [
        11.1391088585,
        14.0372930106,
        7.18088298536,
        58.0335293877,
        40.602210698,
        8.63647600649,
        12.9750533224,
        11.2946067621,
        26.1136650475,
        16.7943855778,
        21.913027435,
        17.1730054506,
        15.6991888392,
        32.5975084656,
        32.3110859828,
        22.331583767,
        52.5387474337,
        8.82102510006,
        39.8392706069,
        12.0900103609,
        17.9166684347,
        34.7415229891,
        54.8329720826,
        35.7410760407,
        122.377063799,
        22.9167105026,
        70.8916095767,
        41.3125784353,
        26.7368108887,
        271.913487877,
        88.7449589977,
        131.066457894,
        93.4001590525,
        145.757910663,
        163.279876724,
        83.9625877264,
        46.6349297363,
        64.2858611097,
        106.77140295,
        94.7749833121,
        75.5204773079,
        11.3915619243,
        10.1128107558,
        8.62819900521,
        10.4939780524,
        23.7495304366,
        13.2163659128,
        6.9721825943,
        7.77384438488,
        10.0141916804,
        5.08846607145,
        17.4941437015,
        5.93246797554,
        7.05183278919,
        55.763858898,
        52.906657773,
        25.451452389,
        51.1906341735,
        23.4798588537,
        7.21354766252,
        9.53414770108,
        14.3374460441,
        9.8043909526,
        9.85053950073,
        4.75222038207,
        9.31474970037,
        7.63782346835,
        10.6798016555,
        11.2470269538,
        11.9803367165,
        9.63515979046,
        22.9579533702,
        14.5329733287,
        7.17064694454,
        9.22258704701,
        8.72139161889,
        7.08565182814,
        9.2014695391,
        16.938221518,
        9.63907264729,
        47.7956736708,
        5.7379832598,
        23.6283928777,
        24.0064794142,
        25.5057338911,
        7.77995618682,
        8.30826446628,
        8.05613393216,
        20.8116006749,
        5.56365195379,
        8.18415644309,
        20.8445279997,
        39.1005354267,
        49.7156083228,
        17.5086949763,
        9.24722262626,
        32.4882452932,
        19.1440221464,
        32.17596964,
        10.7145025149,
        10.0941822543,
        8.40537788854,
        42.0254744233,
        17.3241734412,
        32.3939812877,
        7.68930484181,
        29.8110551293,
        18.5707630735,
        9.07060320447,
        39.1753693725,
        39.925684783,
        12.5406104887,
        64.1079745646,
        36.5620443159,
        40.5074270272,
        15.6793404212,
        32.3634572888,
        16.1819749343,
        8.98043991969,
        60.41114236,
        35.1813737748,
        27.7856152432,
        43.019194651,
        27.5709636544,
        25.3993848571,
        116.724378101,
        43.6317513742,
        16.361305991,
        10.3280259977,
        10.1361214858,
        73.8726146829,
        21.0321102012,
        84.1920947696,
        57.0845625331,
        11.8737800581,
        48.4736644465,
        84.3470333988,
        49.9926852349,
        41.1780352845,
        27.6088754686,
        9.15529505827,
        39.742735469,
        30.266270701,
        51.5500831476,
        28.0831438255,
        45.9434458039,
        40.7338400596,
        31.4728399974,
        48.9491819535,
        28.9345506799,
        20.277667923,
      ],
      "startTime": 2000,
      "type": "FK_SPECTRA",
    },
    "slownessWf": {
      "endTime": 2302,
      "sampleCount": 151,
      "sampleRateHz": 0.5,
      "samples": [
        16.4012194669,
        22.803508502,
        22.803508502,
        13.6014705087,
        15.5241746963,
        12.2065556157,
        1.41421356237,
        16.2788205961,
        11.1803398875,
        5,
        18.0277563773,
        13.152946438,
        21.8403296678,
        14.4222051019,
        18.7882942281,
        1,
        23.0217288664,
        26.9072480941,
        18.7882942281,
        25.179356624,
        19.7230829233,
        17.0293863659,
        14.3178210633,
        13.152946438,
        15.2315462117,
        17.2626765016,
        7.28010988928,
        6.40312423743,
        12.6491106407,
        4.12310562562,
        4,
        3.16227766017,
        4.472135955,
        4.12310562562,
        3,
        3,
        3.16227766017,
        3.16227766017,
        3.16227766017,
        4.472135955,
        3.60555127546,
        1.41421356237,
        5.65685424949,
        15.6204993518,
        2.2360679775,
        24.2074368738,
        20.5182845287,
        7.28010988928,
        21.400934559,
        7.28010988928,
        8.94427191,
        6,
        23.6008474424,
        17.2046505341,
        5.09901951359,
        1.41421356237,
        1.41421356237,
        1.41421356237,
        2.2360679775,
        25.2982212813,
        6.32455532034,
        8.24621125124,
        24.6981780705,
        8.24621125124,
        31.384709653,
        22.360679775,
        7.81024967591,
        7,
        21.4709105536,
        14.4222051019,
        12.3693168769,
        16.4924225025,
        27.3130005675,
        15.6524758425,
        17.88854382,
        24.0831891576,
        23.0867927612,
        19.2353840617,
        8.0622577483,
        24,
        10.7703296143,
        8,
        8.24621125124,
        7.28010988928,
        10.0498756211,
        7,
        15.5241746963,
        1.41421356237,
        5.09901951359,
        9.21954445729,
        6.7082039325,
        4.12310562562,
        3.16227766017,
        14.4222051019,
        11.1803398875,
        12.5299640861,
        23.3452350599,
        3,
        11.6619037897,
        3,
        2.2360679775,
        21.0237960416,
        13.152946438,
        6.40312423743,
        7.81024967591,
        28.7923600978,
        8.60232526704,
        22.8473193176,
        18.6815416923,
        14.3178210633,
        15.2970585408,
        10.1980390272,
        8,
        8.24621125124,
        12.3693168769,
        21.2602916255,
        7.21110255093,
        3.60555127546,
        8.94427191,
        11.1803398875,
        13.6014705087,
        8.0622577483,
        12.3693168769,
        8.24621125124,
        7.21110255093,
        7.28010988928,
        7.07106781187,
        3.16227766017,
        7.61577310586,
        7.07106781187,
        6.40312423743,
        7.61577310586,
        4,
        5,
        4.12310562562,
        8.94427191,
        8.24621125124,
        5.09901951359,
        8.0622577483,
        8.60232526704,
        26.4007575649,
        10.0498756211,
        7.21110255093,
        7.28010988928,
        8.60232526704,
        9.48683298051,
        5.09901951359,
        9.21954445729,
        11.6619037897,
        18.3575597507,
        7.61577310586,
      ],
      "startTime": 2000,
      "type": "FK_SPECTRA",
    },
  },
  "highFrequency": 6,
  "lowFrequency": 3,
  "metadata": {
    "phaseType": "P",
    "slowDeltaX": 111.19492664455875,
    "slowDeltaY": 111.19492664455875,
    "slowStartX": -4447.79706578235,
    "slowStartY": -4447.79706578235,
  },
  "reviewed": false,
  "sampleCount": 151,
  "sampleRateHz": 0.5,
  "startTime": 2000,
  "stepSize": 5,
  "type": "FK_SPECTRA",
  "values": [],
  "windowLead": 1,
  "windowLength": 4,
}
`;

exports[`Update Fk after computeFkSpectra call update Fk COI can handle undefined 1`] = `undefined`;
